packages/react-native-bridge/android/src/main/java/org/wordpress/mobile/WPAndroidGlue/DeferredEventEmitter.java

```diff
@@ -205,6 +210,12 @@ public void onMediaCollectionSaveResult(String firstMediaIdInCollection, boolean
         }
     }
 
+    public void featuredImageIdNotifier(int mediaId) {
```

Comment:
Similar to my naming comment above, what do you think about renaming to `featuredImageIdChanged()` or better:`onFeaturedImageIdChange` to make it read like an event handler? The original `featuredImageIdNotifier` makes me wonder what gets notified... like, is it the user that receives a notification when this function is called?

Reply:
Reading more code where this function is used, I see that it's also called when the JS side [just asks for an update of the featured image id](https://github.com/wordpress-mobile/WordPress-Android/blob/2700d6c49232e6866abdc796cf7cefb5f8c26dea/libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/gutenberg/GutenbergEditorFragment.java#L296). So, the function's purpose is not only to notify the JS side that the id has changed, which means that the event name ([defined here](https://github.com/WordPress/gutenberg/pull/28854/files#diff-bd64b4a30ececb510793aeabfe0b68175eaa825a9f121d8dd953e688a355c01fR46)) is not actually accurate 🤔.

So, if I understand things properly, the event is all in all just trying to pass the the media id of the current featured image, right, whether it was recently updated or not? If so, what do you think about naming the event with `EVENT_FEATURED_IMAGE_ID_CURRENT = "featuredImageIdCurrent"`?

Reply:
> The original featuredImageIdNotifier makes me wonder what gets notified... like, is it the user that receives a notification when this function is called?

This is useful feedback, thank you @hypest! I looked through other similar functions and renamed `featuredImageIdNotifier` to `onRequestFeaturedImageId` in order to hopefully be clearer of the function's purpose. Let me know if you think this is still unclear!

> So, if I understand things properly, the event is all in all just trying to pass the the media id of the current featured image, right, whether it was recently updated or not? If so, what do you think about naming the event with EVENT_FEATURED_IMAGE_ID_CURRENT = "featuredImageIdCurrent"?

That makes sense! The purpose changed as the code evolved (it was originally only passing the id when the image changed) which is why I think the naming became a bit messy/confusing. This is a good lesson to keep on top and iterating on names. Thank you! I've updated this to "featuredImageIdCurrent" now.

Reply:
> I looked through other similar functions and renamed featuredImageIdNotifier to onRequestFeaturedImageId in order to hopefully be clearer of the function's purpose. Let me know if you think this is still unclear!

Aha.

Still trying to grok what the code is trying to do, please bear with me :). There are two rather different code paths that use this function:
1. The JS side asks the native side to provide the featured image id ([getter named function here](https://github.com/WordPress/gutenberg/blob/46ba7e8d770cc0cc1a93664047fdcaaa12f0f0f0/packages/react-native-bridge/android/src/main/java/org/wordpress/mobile/ReactNativeGutenbergBridge/RNReactNativeGutenbergBridgeModule.java#L237-L239)). The native side then [calls `onRequestFeaturedImageId` only if the media id provided matches the featured image id on file](https://github.com/wordpress-mobile/WordPress-Android/blob/622dfad0700d288349fd344783b3bbf24e242cdc/libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/gutenberg/GutenbergEditorFragment.java#L349-L353). By the way, looks like it's not a getter after all since the response to the original `getFeaturedImageId` is conditional. Looks to me more like a "check for something", not a plain getter. If so, would it make sense to rename getFeaturedImageId -> checkIfFeaturedImage?
2. The native side [sends the id when the featured image is set](https://github.com/wordpress-mobile/WordPress-Android/blob/622dfad0700d288349fd344783b3bbf24e242cdc/WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java#L2554).

From flow No2 above, and the function implementation itself ([native side](https://github.com/WordPress/gutenberg/blob/d8c0824c087e5fc76e955123665bc9f2afc893e8/packages/react-native-bridge/android/src/main/java/org/wordpress/mobile/WPAndroidGlue/DeferredEventEmitter.java#L214-L216), https://github.com/WordPress/gutenberg/blob/75de59fd6cacd8faeb787e20c8bc218c24992f46/packages/block-library/src/image/edit.native.js#L305-L319) I understand that onRequestFeaturedImageId just sends the current featured image id to the JS side so, may I propose renaming it to `sendToJSFeaturedImageId`?

Then, flow No1 shows to me we need to better name and document why `onGetFeaturedImageId` doesn't always call sendToJSFeaturedImageId.
Does this all make sense? Sorry if I'm a bit confused on this one.

Reply:
@hypest, thank you again for your help here!  

> By the way, looks like it's not a getter after all since the response to the original getFeaturedImageId is conditional. Looks to me more like a "check for something", not a plain getter. If so, would it make sense to rename getFeaturedImageId -> checkIfFeaturedImage?
>
> Then, flow No1 shows to me we need to better name and document why onGetFeaturedImageId doesn't always call sendToJSFeaturedImageId.

It's right that the `getFeaturedImageId` function is only checking whether an image is featured, I've changed its name to `checkIfFeaturedImage` for the time being.

To further clarify: The `checkIfFeaturedImage` function is called within `componentWillUnmount()` on the JS side. It's currently responsible for flagging which image is featured when the editor is first loaded.

The background for this is that I was unable to get `getEditedPostAttribute` to work to get the post's initial featured image (discussed on project thread and in project's Slack channel). 

I've spent a bit of time with this but am not sure if there's a more elegant way to achieve the same effect, but it's entirely likely I've been looking at the code too long! I'd be interested to hear whether you think this approach makes sense, with those extra details, or if I should work on this a bit more.

Reply:
> To further clarify: The checkIfFeaturedImage function is called within componentWillUnmount() on the JS side. It's currently responsible for flagging which image is featured when the editor is first loaded.

Ah, I see 👍, I will check this out now that I read the logic better. Just to make sure, do you perhaps meant to say `componentDidMount` since that's the one called on editor start?

> I've spent a bit of time with this but am not sure if there's a more elegant way to achieve the same effect, but it's entirely likely I've been looking at the code too long! I'd be interested to hear whether you think this approach makes sense, with those extra details, or if I should work on this a bit more.

Will have a look from than angle and let you know! Can you update the WPAndroid side as well as I think it doesn't compile currently due to bridge changes, thanks! ([Commented on the WPAndroid PR as well](https://github.com/wordpress-mobile/WordPress-Android/pull/14052#issuecomment-810152596))

Reply:
> Just to make sure, do you perhaps meant to say componentDidMount since that's the one called on editor start?

Ah, yes, typo sorry! 🤦‍♀️ I meant `componentDidMount`.

> Can you update the WPAndroid side as well as I think it doesn't compile currently due to bridge changes, thanks!

Updated! :)

Reply:
> Will have a look from than angle and let you know! Can you update the WPAndroid side as well as I think it doesn't compile currently due to bridge changes, thanks! (Commented on the WPAndroid PR as well)

Just an update. I had a short look and I feel like triggering a bridge method to check each image in the post is too much and it might not scale well in image heavy posts. Feel to me that we should pass the initial value of the featured image id during editor boot, find a way to keep that in the global(?) state and check against that for every image block.

To that end, I'm having another look to the original idea around `getEditedPostAttribute` to see if we could hook at that place, to put the id in the global post object, for example. I'm not familiar with that part of the store though so, will continue exploring into it.

Reply:
Did a quick dive and I think that adding a `featured_media: props.featuredMedia` in [the post object we prepare when starting the editor](https://github.com/WordPress/gutenberg/blob/trunk/packages/edit-post/src/editor.native.js#L122-L136), accompanied by passing that `featuredMedia` id [down the bridge](https://github.com/WordPress/gutenberg/blob/trunk/packages/react-native-bridge/android/react-native-bridge/src/main/java/org/wordpress/mobile/WPAndroidGlue/GutenbergProps.kt), is then making `featured_media` available to the `getEditedPostAttribute` Store function. Not sure if that's something you already had tried recently @SiobhyB , let me know if it helps.

That's something to help with the bootstrapping of the editor with the featured image id becoming available then, but will need some more treatment to enable _updating_ the id when the featured image gets set after the editor has already started.

Reply:
@hypest, thank you for this! 🙇‍♀️ I'm working on getting this to work as expected now. I've committed my latest work in progress, which is not yet working, here: https://github.com/WordPress/gutenberg/pull/28854/commits/0cdeadb1ddf8cbc4c491e2eb21b42eda683ac247

I also made the assumption that this will require some changes to the Android side, and have made some initial changes here: https://github.com/wordpress-mobile/WordPress-Android/pull/14052/commits/0ec24df06d5ff419e0f7a21849c1e53afdbf70a8

Does this seem like I'm on the right path? I'm happy to continue troubleshooting and trying to figure this out, but just wanted to make sure I wasn't going off in the wrong direction.

Reply:
Oh, glad you liked the proposal and started going down that route!

Looks like a promising path so far so yeah, please feel free to go ahead. A disclaimer here though: I haven't really worked on getting a real feel of how the _updating_ part will work this time so, please keep that in mind in case this new route proves impractical too after all. Thanks!

Reply:
I've been able to get this working to grab the post's initial featured image ID now and have made use of `setAttributes` to get this working in ComponentDidUpdate [here](https://github.com/WordPress/gutenberg/pull/28854/files#diff-7126ae03000a022bcefa0d2a5fb03b6aca56ed4047e3a49d2131babf229cf6c1R197). How does this seem? Looking forward to any feedback or ideas on this one. :)

Reply:
> How does this seem?

Yeah, I think we're on a good route here. We chatted a bit over Zoom, to share my thoughts on how the the ImageEdit component just needs to listen to the store updates and compare the featured image id it gets from there with the id it keeps in the block attributes, and use those to compute a "isFeaturedImage" boolean inside the render() function.

Then, it becomes a matter of figuring out how to also do the update of the id in the store, for which we can use the [editEntityRecord()](https://github.com/WordPress/gutenberg/blob/trunk/packages/core-data/src/actions.js#L248) action creator (acquiring it via the `withDispach`. But, that part (to update the featured image id value when the image changes) can be done in a separate PR I think.

Let me know what you think!

Reply:
Thank you again for this! I've spent a bit of time today trying to process this and have a few questions:

* What are the overall benefits to moving the `setFeaturedImageState()` functionality inside the `render()` function vs. the current approach? 
* `setFeaturedImageState()` is currently called in `componentDidMount()` for when the editor first loads and then `componentDidUpdate()` for when an image is replaced directly within the image block (not the Post Settings). Will moving the functionality for setting state from `setFeaturedImageState()` to `render()` resolve the need to do that? Is that one of the benefits?
* It seems like I'll need to figure out how to use `editEntityRecord()` to grab the post's current featured image ID _before_ I can move the functionality for setting state from  `setFeaturedImageState()`. Is that right?

Sorry for my fuzziness on this one, just want to make sure I'm clear before moving ahead!

Reply:
It seems all I had to do was ask those questions and then play around a little more to figure out the answers. :) I've been able to convert the functionality previously included in `setFeaturedImageState` to an `isFeaturedImage` and the code is a lot cleaner as a result. 🎉 I've committed those changes.

Reply:
Nice to see the simplification of `isFeaturedImage`!

> setting an isFeaturedImage state

Just a quick comment on the terminology used in the [commit message](https://github.com/WordPress/gutenberg/pull/28854/commits/39303e4a13dce66bfeed96060e452af72e65fb53): now that `isFeaturedImage` is just a helper local const in `render()`, referring to it as "state" can be misleading, mainly because within React, "state" has a special and specific [meaning](https://reactjs.org/docs/state-and-lifecycle.html). I'd recommend referring to it as "local variable/const" to avoid confusion with React's use.